### PR TITLE
Split compile and render into two steps: using intermediate AST tree

### PR DESCRIPTION
**Pull request by zhangciwu**

### CODE REVIEW

File: src/createRenderer.js [hunk 1]
```diff
@@ -80,6 +41,7 @@ export default function createRenderer(tracker, options, overrides = {}) {
     let inlineContent = null;
 
     const elementType = options.elements && options.elements[type];
+    console.log('addElement', tag, props, children, type);
```
Comment (Stupidism) [inline src/createRenderer.js]: debug info?

File: src/compileToIntermediateTree.test.js [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+| ------------- |:-------------:| -----:|
+| col 3 is      | right-aligned | $1600 |
+| col 2 is      | centered      |   $12 |
+| zebra stripes | are neat      |    $1 |
+  `);
+
+  expect(compiled).toMatchSnapshot();
+});
+
+it('should be able to compile codespans', () => {
```
Comment (Stupidism) [inline src/compileToIntermediateTree.test.js:107]: snapshot missing

File: src/__snapshots__/compileToIntermediateTree.test.js.snap [hunk 3]
```diff
@@ -0,0 +1,750 @@
...
+    ],
+  ],
+}
+`;
+
+exports[`should be able to compile image 1`] = `
+Object {
+  "tree": Array [
+    Array [
+      "p",
```
Comment (Stupidism) [inline src/__snapshots__/compileToIntermediateTree.test.js.snap]: Why there's a `p` tag wrapper for image?

File: src/compileToIntermediateTree.test.js [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+
+it('should be able to compile multiple html', () => {
+  const compiled = compileToIntermediateTree(
+    '<div>hello</div>\n<strong>there</strong>\n<em>world</em>'
+  );
+
+  expect(compiled).toMatchSnapshot();
+});
+
+it('should be able to compile components using H and marksy language', () => {
```
Comment (Stupidism) [inline src/compileToIntermediateTree.test.js:133]: h?

File: src/compileToIntermediateTree.test.js [hunk 5]
```diff
@@ -0,0 +1,260 @@
+/* eslint-disable react/destructuring-assignment */
+/* eslint-disable react/jsx-one-expression-per-line */
+import React, { Component, createElement } from 'react';
+import PropTypes from 'prop-types';
+import Preact from 'preact';
+import preactRenderToString from 'preact-render-to-string';
+import { renderToString as infernoRenderToString } from 'inferno-server';
```
Comment (Stupidism) [inline src/compileToIntermediateTree.test.js]: What's inferno?

File: src/compileToIntermediateTree.test.js [hunk 6]
```diff
@@ -0,0 +1,260 @@
+/* eslint-disable react/destructuring-assignment */
+/* eslint-disable react/jsx-one-expression-per-line */
+import React, { Component, createElement } from 'react';
+import PropTypes from 'prop-types';
+import Preact from 'preact';
```
Comment (Stupidism) [inline src/compileToIntermediateTree.test.js]: Why Preact?

File: src/compileToIntermediateTree.test.js [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+
+it('should work with Inferno', () => {
+  const compiled = compileToIntermediateTree(`
+# foo
+  `);
+
+  expect(infernoRenderToString(infernoCreateElement('div', null, compiled.tree))).toMatchSnapshot();
+});
+
+it('should allow injecting context to elements', () => {
```
Comment (Stupidism) [inline src/compileToIntermediateTree.test.js:176]: What does this test case intend to test?

File: src/compileToIntermediateTree.test.js [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+    {},
+    {
+      foo: 'bar',
+    }
+  );
+
+  expect(compiled).toMatchSnapshot();
+});
+
+it('should allow overriding inline code element', () => {
```
Comment (Stupidism) [inline src/compileToIntermediateTree.test.js]: snapshot missing

File: src/compileToIntermediateTree.test.js [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+  expect(compiled).toMatchSnapshot();
+});
+
+it('should allow overriding block code element', () => {
+  const compiled = compileToIntermediateTree('```js\ncode\n```');
+
+  expect(compiled).toMatchSnapshot();
+});
+
+it('should escape code when no highlighting is supplied', () => {
```
Comment (Stupidism) [inline src/compileToIntermediateTree.test.js]: No highlighting supplied?
> ```\nconst Foo = () => <div/>\n```

File: src/compileToIntermediateTree.test.js [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+  const compiled = compileToIntermediateTree('```js\nconst Foo = () => <div/>\n```');
+
+  expect(compiled).toMatchSnapshot();
+});
+
+it('should highlight code with highlight.js', () => {
```
Comment (Stupidism) [inline src/compileToIntermediateTree.test.js]: How does this test case work?

File: src/codeRenderer.js [hunk 11]
```diff
@@ -0,0 +1,39 @@
+export const CodeComponent = options => props => {
```
Comment (Stupidism) [inline src/codeRenderer.js:1]: Do we even need to support code?

**Review by Stupidism — CHANGES_REQUESTED** [review]

File: src/createRenderer.js (hunk 1 above)
Comment (zhangciwu) [reply src/createRenderer.js]: removed

File: src/codeRenderer.js (hunk 11 above)
Comment (zhangciwu) [reply src/codeRenderer.js:1]: yes

File: src/compileToIntermediateTree.test.js (hunk 10 above)
Comment (zhangciwu) [reply src/compileToIntermediateTree.test.js]: removed

File: src/compileToIntermediateTree.test.js (hunk 9 above)
Comment (zhangciwu) [reply src/compileToIntermediateTree.test.js]: nope

File: src/compileToIntermediateTree.test.js (hunk 8 above)
Comment (zhangciwu) [reply src/compileToIntermediateTree.test.js]: added

File: src/compileToIntermediateTree.test.js (hunk 7 above)
Comment (zhangciwu) [reply src/compileToIntermediateTree.test.js:176]: context injection

File: src/compileToIntermediateTree.test.js (hunk 5 above)
Comment (zhangciwu) [reply src/compileToIntermediateTree.test.js]: something like preact?

File: src/compileToIntermediateTree.test.js (hunk 4 above)
Comment (zhangciwu) [reply src/compileToIntermediateTree.test.js:133]: `h(Component)`

File: src/__snapshots__/compileToIntermediateTree.test.js.snap (hunk 3 above)
Comment (zhangciwu) [reply src/__snapshots__/compileToIntermediateTree.test.js.snap]: wrapper

**Review by Stupidism — APPROVED** [review]

**Pull request merged**